astroq/io.py

```diff
@@ -75,6 +75,115 @@ def serialize_schedule(Yrds, semester_planner):
     build_fullness_report(semester_planner, "Round1")
     return sparse
 
+def compute_program_statistics(semester_planner, schedule_df):
+    """
+    Compute awarded, requested, and scheduled statistics per program.
+    
+    Args:
+        semester_planner (obj): a SemesterPlanner object from splan.py
+        schedule_df (DataFrame): DataFrame containing scheduled observations
+        
+    Returns:
+        dict: Dictionary keyed by program name, with values containing:
+            - awarded_nights, awarded_hours, awarded_slots
+            - past_nights, past_hours, past_slots
+            - requested_nights, requested_hours, requested_slots
+            - scheduled_nights, scheduled_hours, scheduled_slots
+            - fullness1 ((scheduled+past)/requested %)
+            - fullness2 ((scheduled+past)/awarded %)
+    """
+    # Get conversion factors
+    slot_size = semester_planner.slot_size  # in minutes
+    hours_per_night = semester_planner.hours_per_night  # in hours
+    slots_per_hour = 60 / slot_size
+    slots_per_night = hours_per_night * slots_per_hour
+    
+    # Read programs.csv to get awarded nights per program
+    program_frame = pd.read_csv(semester_planner.programs_file)
+    
+    # Initialize result dictionary
+    program_stats = {}
+    
+    # Process each program
+    for _, prog_row in program_frame.iterrows():
```

Comment:
These summary stats should be easily computable in a few lines with schedule_df.groupby() and applying aggregating functions